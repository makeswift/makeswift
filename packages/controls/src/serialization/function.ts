export type AnyFunction = (...args: any) => any

export function isFunction(value: any): value is AnyFunction {
  return typeof value === 'function'
}

export type SerializedFunctionReturnType<T extends AnyFunction> = Awaited<
  ReturnType<T>
>

export type DeserializedFunction<T extends AnyFunction> = (
  ...args: Parameters<T>
) => Promise<SerializedFunctionReturnType<T>>
