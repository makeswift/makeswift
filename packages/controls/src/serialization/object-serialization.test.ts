import { AnyFunction } from './function'
import {
  deserializeObject,
  SerializationPlugin,
  serializeObject,
  type Deserialized,
} from './object-serialization'

const mockSerializeFunctionPlugin: SerializationPlugin<AnyFunction> = {
  match: (val: unknown) => typeof val === 'function',
  serialize: jest.fn(() => `[SerializedFunction]`),
}

const mockSerializeSerializableObjectPlugin: SerializationPlugin<SerializableObject> =
  {
    match: (val: unknown) => val instanceof SerializableObject,
    serialize: jest.fn((def) => def.serialize()),
  }

class SerializableObject {
  constructor(private readonly obj: { type: string; [key: string]: any }) {}

  serialize(): unknown {
    return serializeObject(this.obj, [
      mockSerializeFunctionPlugin,
      mockSerializeSerializableObjectPlugin,
    ])
  }
}

afterEach(() => {
  jest.clearAllMocks()
})

describe('object serialization', () => {
  test('Deserialized', () => {
    const obj = {
      label: 'hello',
      getItemLabel: (item: { label: string }) => item.label,
    }

    const deserialized: Deserialized<typeof obj> = {
      label: 'hello',
      getItemLabel: async (item: { label: string }) => item.label,
    }

    expect(deserialized.getItemLabel).toMatchSnapshot()
  })

  describe.each([
    {},
    { label: undefined },
    { label: 'visible', defaultValue: true },
    { label: 'visible', options: [{ label: 'one', value: 1 }] },
  ])('%o', (object) => {
    test('should serialize to itself', () => {
      expect(serializeObject(object)).toEqual(object)
    })

    test('should deserialize to itself', () => {
      const serialized = serializeObject(object)
      expect(deserializeObject(serialized)).toEqual(object)
    })
  })

  test('should correctly serialize nested Serializables', () => {
    const object = {
      item: new SerializableObject({
        type: 'item',
        getItemLabel: (item: any) => item.label,
        subitems: [
          new SerializableObject({
            type: 'subitem 1',
            arrayOfFunctions: [(item: any) => item.label],
          }),
          new SerializableObject({
            type: 'subitem 2',
            getItemLabel: (item: any) => item.label,
          }),
        ],
      }),
    }

    const serialized = serializeObject(object, [
      mockSerializeFunctionPlugin,
      mockSerializeSerializableObjectPlugin,
    ])
    expect(mockSerializeFunctionPlugin.serialize).toHaveBeenCalledTimes(3)
    expect(serialized).toMatchSnapshot()
  })

  describe.each([
    { getItemLabel: (item: any) => item.label },
    { arrayOfFunctions: [(item: any) => item.label] },
  ])('%o', (object) => {
    test('should handle function serialization', async () => {
      const serialized = serializeObject(object, [
        mockSerializeFunctionPlugin,
        mockSerializeSerializableObjectPlugin,
      ])
      expect(serialized).toMatchSnapshot()
    })

    test('should deserialize serialized functions', () => {
      const serialized = serializeObject(object)
      expect(deserializeObject(serialized)).toMatchSnapshot()
    })
  })
})
