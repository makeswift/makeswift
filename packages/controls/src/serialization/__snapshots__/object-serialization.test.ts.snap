// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object serialization {
  arrayOfFunctions: [ [Function (anonymous)] { [length]: 1, [name]: '' }, [length]: 1 ]
} should deserialize serialized functions 1`] = `
{
  "arrayOfFunctions": [
    [Function],
  ],
}
`;

exports[`object serialization {
  arrayOfFunctions: [ [Function (anonymous)] { [length]: 1, [name]: '' }, [length]: 1 ]
} should handle function serialization 1`] = `
{
  "arrayOfFunctions": [
    "[SerializedFunction]",
  ],
}
`;

exports[`object serialization {
  getItemLabel: [Function: getItemLabel] { [length]: 1, [name]: 'getItemLabel' }
} should deserialize serialized functions 1`] = `
{
  "getItemLabel": [Function],
}
`;

exports[`object serialization {
  getItemLabel: [Function: getItemLabel] { [length]: 1, [name]: 'getItemLabel' }
} should handle function serialization 1`] = `
{
  "getItemLabel": "[SerializedFunction]",
}
`;

exports[`object serialization Deserialized 1`] = `[Function]`;

exports[`object serialization should correctly serialize nested Serializables 1`] = `
{
  "item": {
    "getItemLabel": "[SerializedFunction]",
    "subitems": [
      {
        "arrayOfFunctions": [
          "[SerializedFunction]",
        ],
        "type": "subitem 1",
      },
      {
        "getItemLabel": "[SerializedFunction]",
        "type": "subitem 2",
      },
    ],
    "type": "item",
  },
}
`;
